### ISSUE TITLE
Link log entries to a record id

### ISSUE THREAD

**Issue by mikbranchaud** [j-fischer/rflib#66]
I would like to be able to quickly search log for a particular Id, usually a Case Id for my project.  This is somewhat easy in the platform event monitor component (though it can sometimes take a long time for all the entries from the past three days to load into the component).  But if the original issue occurred more than 72 hours in the past and we want to troubleshoot the root cause without reproducing the issue (frequently reproducing the issue has side affects we want to avoid) then we would need to search the archive log in the Message attribute (assuming we consistently put the record id in the message

It would be nice if the LOGGER.<level>() call optionally accepted a record id which was populated in its own field on the archive big object.  I could then query by this column directly.  I could also create a related list (not sure if a custom what id can be configred which would be supported by the new dynamic related list feature but we have a custom related list aura component that will just run any soql to filter a list which we could use to filter by the record page reccord id)

**Comment by j-fischer** [comment]
Hi @mikbranchaud , apologies for the delayed reply, but I had to do some research. 

Basically, what you are suggesting cannot be done, based on my understanding. If you add a record ID field to the big object and want to search for it, [it must be part of the index](Specifies whether the field is required. All fields that are part of the index must be marked as required.). If it is part of the index, it is a required field and must be provided with every request. Even if the framework would put a default value for blanks, I am not convinced yet that this is the best design, to be honest. 

Looking at the scenario you described, you should be able to search for the ID in the log messages. If you would want to add it to the method signatures, I would assume you could add the record ID to a log message as well. Then, when trying to debug and find the record ID in the logs, all you have to do is to search the archive for a wide date range. You can return up to 50,000 records and then search locally. The search is super fast, so there should not be any performance implications. 

My statement works under the assumption that you have a rough idea about the timing of an incident involving a specific record. 

Does this all make sense? I would like to hear some more details about your use case, specifically why the above process does not work for you. I did consider adding a record ID in the past but decided against it because the ID could be added to the log messages and it felt that the method overrides in the logger where not as compatible with standard log method signatures this way, which is one thing I wanted to maintain. 

Now, I am still willing to consider making such an addition, but I would like to discuss this further before making any decisions. 

Thanks

**Comment by j-fischer** [comment]
@mikbranchaud 
I thought about this some more. Even if I was to add a record ID field to the index (and if I remember correct, I would have to create an entire new Big Object because of the package limitations), I would still keep the date fields as index 1 & 2. In that case, you would have to provide a date range even if searching for the record ID. And this scenario is already covered with the current implementation, except that you would search the record ID in the log message. 

To sum things up, I would definitely recommend to add record IDs to the log message, I do that all the time. Search for a wide date range to maximize the number of records. The frontend can easily handle 50k records and search those for the ID within less than 2 seconds. This should cover your use case. 

I hope you are ok if I am closing the ticket. If I missed anything, please feel free to re-open it. 

Thanks,
Johannes

**Comment by mikbranchaud** [comment]
I am trying to use the log archive as is with your recommendation.  I am not seeing events in the archive big object for some reason.  For my user's profile I have set he Archive Log Level to INFO and created a flow which uses the log action in its exception path with a log level of ERROR.  When I run the flow through the fault path I am not seeing any records in the big object either through the monitor LWC or by querying the object in devconsole.  I started looking at the code and other than the invocable action in rflib_ArchiveLogAction, it is not clear to me when the messages are inserted?  Could you elaborate on what triggers an insert to the archive?

**Comment by j-fischer** [comment]
> Looking at the rflib_Log_Event_Handler which invokes the action, this flow is triggered from the platform event creation. So my reading of this is that the log archive must in practice archive at event log levels equal to or less than the platform event log level. In other words I could not use the log archive in lieu of the platform event.

This understanding is absolutely correct. The main driver of the archive is still the platform event. The reason for this is to reduce the impact on the production transaction. The more the framework does beyond the platform event submission, the more it interferes with the original transaction and can be the cause of errors. Now, while big objects are less intrusive, the archive still follows the main design principal of the framework. 

> My impression from the wiki was that the intent of the log archive was to minimize the impact of the logging feature on limits related to platform event but if I need to keep platform event logging relatively high then wouldn't the number of platform events logged and subscribed too remain relatively high?

The purpose of the archive is two fold: First, it allows to store log events for longer than the 72 hours. Second, it helps with management of[ Platform Event Delivery governor limits](https://developer.salesforce.com/docs/atlas.en-us.platform_events.meta/platform_events/platform_event_limits.htm) when using the historic messages connection mode. You are also correct that the intention of the Archive is, if used, to make the historic-messages mode obsolete, so that you can jump right to the archive. 

My question for you would be why you need to log permanently using the INFO log level? This would be very redundant as the log stack would repeatedly capture the same messages in the event. And if there was a WARN (or higher, i.e. ERROR) message, it would contain all your INFO/DEBUG messages. I assign the INFO level only to individuals (not as the global default) in production, or use it in my personal development org where I am the only user. 

> Be able to log entries into the archive without worrying about the platform events. If you think that raising log level to say DEBUG for all users would not have an adverse impact on limits then I suppose I could raise the platform event logging level and just ignore the platform log and query the archive immediately

INFO is the lowest level that you can publish. This is by design to avoid recursions through the trigger framework logs, which are all set to DEBUG for the purpose of not triggering any events, with the exception of a trigger handler throwing an exception. I do not recommend to publish INFO log messages globally in production unless in extreme debugging scenarios for a short amount of time. 

> If the log level is low, say ERROR, still have the messages contain ALL entries for higher log levels (DEBUG, TRACE) for the number of entries prior to the error occurring (currently we have the default 100)

The framework already does this, if you want more messages than the 100, simply increase the stack size in your Logger Settings. 

RFLIB is designed to only log messages in production when something is wrong. Again, this is to create the absolute smallest footprint of Governor Limits to ensure that those are reserved for your production code, not the logging framework. 

Would it help if we jump on a quick call in the near future to discuss this? I currently feel that the framework has all the features for what you want to accomplish, we just may need to tweak the settings. 

I hope all of these answers helped!
